### ISSUE TITLE
Cannot use ExternalExtractedLicenseInfo

### ISSUE THREAD

**Issue by nicoweidner** [spdx/spdx-java-jackson-store#46]
Disclaimer: I don't have a good understanding of the "external" entities `ExternalExtractedLicenseInfo` and `ExternalSpdxElement`, but was rather following the code.

I tried to use a `ExternalExtractedLicenseInfo` in a document, adding it to the list of extractedLicenseInfos, and to serialize that doc to json. Then `doc.verify()` passes, but when trying to serialize to json, the error `Extracted License Infos not of type ExtractedLicensingInfo`. That feels wrong to me, since `ExternalExtractedLicenseInfo extends ExtractedLicenseInfo`.

A halfway complete test case: Here, `doc` is some minimal document. The error is thrown on the last line, when trying to serialize the document.
```
        var externalDocumentRef = doc.createExternalDocumentRef("DocumentRef-1", "externalDocUri", sha1Checksum);
        var externalLicenseInfo = new ExternalExtractedLicenseInfo(doc.getModelStore(), doc.getDocumentUri(), "DocumentRef-1:LicenseRef-XXX", doc.getCopyManager(), true);
        doc.setExternalDocumentRefs(List.of(externalDocumentRef));
        
        doc.setExtractedLicenseInfos(List.of(externalLicenseInfo));
        
        assertThat(doc.verify()).isEmpty();

        var serializer = new JacksonSerializer(new ObjectMapper(),
                MultiFormatStore.Format.JSON_PRETTY, MultiFormatStore.Verbose.COMPACT,
                doc.getModelStore());
        var jsonNode = serializer.docToJsonNode(doc.getDocumentUri());
```

**Comment by goneall** [comment]
@nicoweidner The issue is that the `ExternalExtractedLicenseInfo` is intended to be defined in an external document - not within the document containing the reference.  Think of `ExternalExtractedLicenseInfo` and `ExternalSpdxElement` as pointers to a definition of a license or element _outside_ of the SPDX document itself.  The list of `ExtractedLicenseInfos` in the document is definitions of `ExtractedLicenseInfos` defined within the document.

You don't need to add an `ExternalExtractedLicenseInfo` to the list `ExtractedLicenseInfos` to use it within a document.  In fact, trying to do so should cause an error - which  doesn't today.  We should add a check in the library for this.  There does seem to be a check in the serializer for this - which is OK.

**Comment by nicoweidner** [comment]
Hm... I see, thanks for the explanation. So an intended usage of external license info would be as follows, I suppose? Here, I just create the `ExternalExtractedLicenseInfo` and then reference it in a file.
```
        var externalDocumentRef = doc.createExternalDocumentRef("DocumentRef-1",
                "externalDocUri", sha1Checksum);        
        var externalLicenseInfo = new ExternalExtractedLicenseInfo(doc.getModelStore(),
                doc.getDocumentUri(), "DocumentRef-1:LicenseRef-XXX",
                doc.getCopyManager(), true);
        var file = doc.createSpdxFile("SPDXRef-file", "./foo.txt", null,
                        List.of(), null, sha1Checksum)
                .build();
        file.setLicenseConcluded(externalLicenseInfo);
```
In this case, `licenseConcluded` on the file shows up as `externalDocUri#LicenseRef-XXX`, which looks ok to me.


One issue I have with the current structure is that `ExternalExtractedLicenseInfo` is a subclass of `ExtractedLicenseInfo`. Not being able to use it in `ExtractedLicenseInfos` seems like a violation of the Liskov substitution principle to me (or, in simpler terms, a confusing data structure).

**Comment by goneall** [comment]
> So an intended usage of external license info would be as follows, I suppose?

Yes - that is correct

> One issue I have with the current structure is that `ExternalExtractedLicenseInfo` is a subclass of `ExtractedLicenseInfo`.

I agree it is confusing.  Both `ExternalSpdxElement` and `ExternalExtractedLicenseInfo` are intended to be used referenced  wherever an `SpdxElement` or an `ExtractedLicenseInfo` is allowed.  So, from that perspective they can be "substituted".  Where it gets confusing is that you cannot fetch some of the properties and you can not write any of the properties associated with those data structures, so it is not fully substitutable.  It is similar to having an immutable version of a standard class where you can't write to the properties.

I'm open to suggestions on alternative structures that are more straightforward.  The requirements would be that any property expecting an `Element` or `ExtractractedLicenseInfo` type would accept the `ExternalSpdxElement` and `ExternalExtractedLicenseInfo` typed objects resp.  Since these objects are not contained within the document, we would not be able to access the individual property values with the exception of the ID.

**Comment by goneall** [comment]
@nicoweidner Thanks for the suggestion.  I like the idea of having an abstract parent class.  Why didn't I think of that earlier!

I don't think this would break anything if we make the parent class abstract.

I'll work on a PR and see if it tests out OK.

**Comment by goneall** [comment]
Moving this over to the library repo.
